diff --color -rupN ./a/contents/code/main.js ./b/contents/code/main.js
--- a/contents/code/main.js	2022-05-29 00:00:00.000000000 +0200
+++ b/contents/code/main.js	2022-06-03 23:19:17.825485403 +0200
@@ -1,3 +1,6 @@
-workspace.clientActivated.connect(function(client) {
-    callDBus("com.system76.Scheduler", "/com/system76/Scheduler", "com.system76.Scheduler", "SetForegroundProcess", client.pid);
-})
+workspace.clientAdded.connect(function(client) {
+	if (!client.minimized) {
+	callDBus("org.kde.klauncher5", "/KLauncher", "", "exec_blind", "kshell5", new Array("/etc/system76-scheduler/system76-scheduler-dbus.sh", client.pid)); 
+};
+});
+
diff --color -rupN ./a/put_to_etc_system76_scheduler/system76-scheduler-dbus.sh ./b/put_to_etc_system76_scheduler/system76-scheduler-dbus.sh
--- a/put_to_etc_system76_scheduler/system76-scheduler-dbus.sh	1970-01-01 01:00:00.000000000 +0100
+++ b/put_to_etc_system76_scheduler/system76-scheduler-dbus.sh	2022-06-03 23:22:49.858659050 +0200
@@ -0,0 +1,11 @@
+#!/bin/bash
+# https://github.com/maxiberta/kwin-system76-scheduler-integration
+
+PID=$1
+SERVICE="com.system76.Scheduler"
+PATH="/com/system76/Scheduler"
+INTERFACE="com.system76.Scheduler"
+METHOD="SetForegroundProcess"
+
+
+/usr/bin/qdbus --system $SERVICE $PATH $INTERFACE.$METHOD $PID >/dev/null 2>&1
